File: packages/frontend/types/workerQuestion.ts
/**
 * ワーカー質問関連の型定義
 *
 * バックエンドのワーカー質問スキーマに対応するフロントエンドの型定義
 */

/**
 * 質問ステータス
 */
export enum QuestionStatus {
  PENDING = "pending",
  IN_REVIEW = "in_review",
  ANSWERED = "answered",
  CLOSED = "closed",
}

/**
 * 質問優先度
 */
export enum QuestionPriority {
  LOW = "low",
  MEDIUM = "medium",
  HIGH = "high",
  URGENT = "urgent",
}

/**
 * ワーカー質問情報
 */
export interface WorkerQuestion {
  id: number;
  workerId: number;
  organizationId: number;
  clientOrganizationId: number | null;
  title: string;
  content: string;
  status: QuestionStatus;
  priority: QuestionPriority;
  answer: string | null;
  answeredByUserId: number | null;
  answeredAt: string | null;
  tags: string | null;
  internalNotes: string | null;
  createdAt: string;
  updatedAt: string;
  deletedAt: string | null;
}

/**
 * 質問作成リクエスト
 */
export interface WorkerQuestionCreateRequest {
  title: string;
  content: string;
  clientOrganizationId?: number | null;
  priority?: QuestionPriority;
  tags?: string | null;
}

/**
 * 質問更新リクエスト
 */
export interface WorkerQuestionUpdateRequest {
  title?: string | null;
  content?: string | null;
  status?: QuestionStatus | null;
  priority?: QuestionPriority | null;
  tags?: string | null;
  internalNotes?: string | null;
}

/**
 * 回答追加リクエスト
 */
export interface WorkerQuestionAnswerRequest {
  answer: string;
}

/**
 * 質問一覧レスポンス
 */
export interface WorkerQuestionListResponse {
  questions: WorkerQuestion[];
  total: number;
  skip: number;
  limit: number;
}

/**
 * 未読数レスポンス
 */
export interface UnreadCountResponse {
  unreadCount: number;
}

/**
 * 質問一覧取得パラメータ
 */
export interface WorkerQuestionListParams {
  skip?: number;
  limit?: number;
  status?: QuestionStatus;
  priority?: QuestionPriority;
  workerId?: number;
}

/**
 * 質問ステータスの日本語ラベル
 */
export const QuestionStatusLabel: Record<QuestionStatus, string> = {
  [QuestionStatus.PENDING]: "未対応",
  [QuestionStatus.IN_REVIEW]: "確認中",
  [QuestionStatus.ANSWERED]: "回答済み",
  [QuestionStatus.CLOSED]: "クローズ",
};

/**
 * 質問優先度の日本語ラベル
 */
export const QuestionPriorityLabel: Record<QuestionPriority, string> = {
  [QuestionPriority.LOW]: "低",
  [QuestionPriority.MEDIUM]: "中",
  [QuestionPriority.HIGH]: "高",
  [QuestionPriority.URGENT]: "緊急",
};

/**
 * 質問ステータスの色（Tailwind CSSクラス）
 */
export const QuestionStatusColor: Record<QuestionStatus, string> = {
  [QuestionStatus.PENDING]: "bg-yellow-100 text-yellow-800",
  [QuestionStatus.IN_REVIEW]: "bg-blue-100 text-blue-800",
  [QuestionStatus.ANSWERED]: "bg-green-100 text-green-800",
  [QuestionStatus.CLOSED]: "bg-gray-100 text-gray-800",
};

/**
 * 質問優先度の色（Tailwind CSSクラス）
 */
export const QuestionPriorityColor: Record<QuestionPriority, string> = {
  [QuestionPriority.LOW]: "bg-gray-100 text-gray-800",
  [QuestionPriority.MEDIUM]: "bg-blue-100 text-blue-800",
  [QuestionPriority.HIGH]: "bg-orange-100 text-orange-800",
  [QuestionPriority.URGENT]: "bg-red-100 text-red-800",
};

File: packages/backend/src/application/schemas/worker_question.py
"""
ワーカー質問関連のPydanticスキーマ定義

API境界でのバリデーションとデータ変換を行うDTOスキーマ
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from src.infrastructure.persistence.models.worker_question import QuestionPriority, QuestionStatus


# ========================================
# リクエストスキーマ
# ========================================


class WorkerQuestionCreateRequest(BaseModel):
    """ワーカー質問作成リクエスト"""

    title: str = Field(
        ...,
        min_length=1,
        max_length=500,
        description="質問タイトル",
        examples=["フォーム入力の手順について"],
    )
    content: str = Field(
        ...,
        min_length=1,
        description="質問内容",
        examples=["〇〇フォームの入力手順がわからないため、教えていただけますでしょうか。"],
    )
    client_organization_id: int | None = Field(
        None,
        description="関連する顧客組織ID（オプション）",
        examples=[1],
    )
    priority: QuestionPriority = Field(
        default=QuestionPriority.MEDIUM,
        description="質問優先度",
        examples=[QuestionPriority.MEDIUM],
    )
    tags: str | None = Field(
        None,
        max_length=500,
        description="タグ（JSON配列形式、オプション）",
        examples=['["フォーム入力", "操作方法"]'],
    )


class WorkerQuestionUpdateRequest(BaseModel):
    """ワーカー質問更新リクエスト"""

    title: str | None = Field(
        None,
        min_length=1,
        max_length=500,
        description="質問タイトル",
        examples=["フォーム入力の手順について（修正）"],
    )
    content: str | None = Field(
        None,
        min_length=1,
        description="質問内容",
        examples=["〇〇フォームの〇〇項目の入力手順がわからないため、教えていただけますでしょうか。"],
    )
    status: QuestionStatus | None = Field(
        None,
        description="質問ステータス",
        examples=[QuestionStatus.IN_REVIEW],
    )
    priority: QuestionPriority | None = Field(
        None,
        description="質問優先度",
        examples=[QuestionPriority.HIGH],
    )
    tags: str | None = Field(
        None,
        max_length=500,
        description="タグ（JSON配列形式、オプション）",
        examples=['["フォーム入力", "操作方法", "緊急"]'],
    )
    internal_notes: str | None = Field(
        None,
        max_length=2000,
        description="営業支援会社用の内部メモ",
        examples=["この質問は〇〇案件に関連しています"],
    )


class WorkerQuestionAnswerRequest(BaseModel):
    """ワーカー質問回答リクエスト"""

    answer: str = Field(
        ...,
        min_length=1,
        description="回答内容",
        examples=["〇〇フォームの入力手順は以下の通りです。\n1. ・・・\n2. ・・・"],
    )


# ========================================
# レスポンススキーマ
# ========================================


class WorkerQuestionResponse(BaseModel):
    """ワーカー質問レスポンス"""

    id: int = Field(..., description="質問ID")
    worker_id: int = Field(..., description="質問者のワーカーID")
    organization_id: int = Field(..., description="営業支援会社の組織ID")
    client_organization_id: int | None = Field(None, description="関連する顧客組織ID")
    title: str = Field(..., description="質問タイトル")
    content: str = Field(..., description="質問内容")
    status: QuestionStatus = Field(..., description="質問ステータス")
    priority: QuestionPriority = Field(..., description="質問優先度")
    answer: str | None = Field(None, description="回答内容")
    answered_by_user_id: int | None = Field(None, description="回答者のユーザーID")
    answered_at: datetime | None = Field(None, description="回答日時")
    tags: str | None = Field(None, description="タグ（JSON配列形式）")
    internal_notes: str | None = Field(None, description="営業支援会社用の内部メモ")
    created_at: datetime = Field(..., description="作成日時")
    updated_at: datetime = Field(..., description="更新日時")
    deleted_at: datetime | None = Field(None, description="削除日時（論理削除）")

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": 1,
                "worker_id": 1,
                "organization_id": 1,
                "client_organization_id": 1,
                "title": "フォーム入力の手順について",
                "content": "〇〇フォームの入力手順がわからないため、教えていただけますでしょうか。",
                "status": "pending",
                "priority": "medium",
                "answer": None,
                "answered_by_user_id": None,
                "answered_at": None,
                "tags": '["フォーム入力", "操作方法"]',
                "internal_notes": None,
                "created_at": "2025-11-22T10:00:00Z",
                "updated_at": "2025-11-22T10:00:00Z",
                "deleted_at": None,
            }
        },
    )


class WorkerQuestionListResponse(BaseModel):
    """ワーカー質問一覧レスポンス"""

    questions: list[WorkerQuestionResponse] = Field(..., description="質問リスト")
    total: int = Field(..., description="総件数")
    skip: int = Field(..., description="スキップした件数")
    limit: int = Field(..., description="取得した件数")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "questions": [
                    {
                        "id": 1,
                        "worker_id": 1,
                        "organization_id": 1,
                        "client_organization_id": 1,
                        "title": "フォーム入力の手順について",
                        "content": "〇〇フォームの入力手順がわからないため、教えていただけますでしょうか。",
                        "status": "pending",
                        "priority": "medium",
                        "answer": None,
                        "answered_by_user_id": None,
                        "answered_at": None,
                        "tags": '["フォーム入力", "操作方法"]',
                        "internal_notes": None,
                        "created_at": "2025-11-22T10:00:00Z",
                        "updated_at": "2025-11-22T10:00:00Z",
                        "deleted_at": None,
                    }
                ],
                "total": 1,
                "skip": 0,
                "limit": 100,
            }
        },
    )


class UnreadCountResponse(BaseModel):
    """未読質問数レスポンス"""

    unread_count: int = Field(..., description="未読質問数（pendingステータスの質問数）")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "unread_count": 5,
            }
        },
    )
